### PR TITLE
Add SerializeOnlyNonEmpty

### PR DESCRIPTION
**Pull request by vanniktech**
Still WIP:

- [X] what should we do about primitive arrays? extend instance of check?
- [X] toString still fails

Do you have any better idea than using `instanceof`?

There is `Class<?>.isArray()` but AFAIK not a way to see whether it's empty or not.

### CODE REVIEW

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  @Override public T fromJson(JsonReader reader) throws IOException {
+    return delegate.fromJson(reader);
+  }
+
+  @Override public void toJson(JsonWriter writer, T value) throws IOException {
+    if (isNotEmpty(value)) {
+      delegate.toJson(writer, value);
+    } else {
+      // We'll need to consume this property otherwise we'll get an IllegalArgumentException.
+      delegate.toJson(writer, null);
```
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: call `writer.setSerializeNulls(true)` before consuming the value and reset to previous config. Check out `SerializeNulls` for reference.

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  }
+
+  private boolean isNotEmpty(final T value) {
+    if (value instanceof Object[]) {
+      Object[] array = (Object[]) value;
+      return array.length > 0;
```
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: you can save a flag (`isArray` for example) and just cast to `T[]` then primitive support is not an issue.

**Comment by serj-lotutovici** [comment]
The `toString` test fails, because moshi's `array` adapter does not have a custom toString. Maybe makes more sense to test if the returned value ends with out custom postfix.

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 2 above)
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: You can't cast `T` to `T[]` in the case of `int[]`

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 1 above)
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: I'd argue that this should not be done here. I only want to serialize non empty if I want to serialize non empty + null, I'd apply the `@SerializeOnlyNonEmpty` & `@SerializeNulls` annotation.

Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: aren't they kinda exclusive? if you don't want to serialise a non empty array, I'd assume that is should not be serialised under any circumstances.

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 2 above)
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: hm... that's not good.

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 1 above)
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: @SerializeOnlyNonEmpty null  -> don't serialize
@SerializeOnlyNonEmpty []     -> don't serialize
@SerializeOnlyNonEmpty [5]   -> serialize

If you wanted to serialize the first one you can add `@SerializeNulls`

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 2 above)
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: Yup I'll add the instanceof checks then and add plenty of tests

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    } else if (value instanceof Collection) {
+      Collection collection = (Collection) value;
+      return collection.size() > 0;
+    } else if (value instanceof Map) {
+      Map map = (Map) value;
```
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: @serj-lotutovici you want me to inline all of those? wasn't sure which way you prefer

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 2 above)
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: I think a found a slightly better way to do this:
```
if (isArray) {
      return value != null && Array.getLength(value) > 0;
}
```

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 3 above)
Comment (serj-lotutovici) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: As long as it's readable, it's ok.

File: src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java (hunk 2 above)
Comment (vanniktech) [inline src/main/java/com/serjltt/moshi/adapters/SerializeOnlyNonEmptyJsonAdapter.java]: That's awesome !

**Pull request merged**